File: sdks/lua/test/test_ops.lua
--- Tests for Client:hello/put/put_null/introspect/resolve/eval/watch, and
--- for the unsupported-capability path (a backend that omits an optional
--- method) — against a mock backend, matching test_client.lua's pattern.

return function(suite, test, skip, assert_eq, assert_true, assert_nil, assert_not_nil)
  local client_mod  = require("beachcomber.client")
  local Client      = client_mod.Client

  local function make_backend(overrides)
    local backend = { name = function(_) return "mock" end }
    for k, v in pairs(overrides or {}) do backend[k] = v end
    return backend
  end

  -- ── hello ─────────────────────────────────────────────────────────────────

  suite("Client:hello")

  test("hello returns protocol_version and daemon_version", function()
    local backend = make_backend({
      hello = function() return { ok = true, data = { protocol_version = "1", daemon_version = "0.9.0" } } end,
    })
    local info, err = Client.new(backend):hello()
    assert_nil(err)
    assert_eq(info.protocol_version, "1")
    assert_eq(info.daemon_version, "0.9.0")
  end)

  test("hello defaults empty strings when data missing", function()
    local backend = make_backend({ hello = function() return { ok = true } end })
    local info, err = Client.new(backend):hello()
    assert_nil(err)
    assert_eq(info.protocol_version, "")
    assert_eq(info.daemon_version, "")
  end)

  test("hello returns nil+Error on backend error", function()
    local backend = make_backend({
      hello = function() return { ok = false, error = { kind = "server_error", message = "not implemented" } } end,
    })
    local info, err = Client.new(backend):hello()
    assert_nil(info)
    assert_not_nil(err)
  end)

  test("hello returns Error.unsupported when the backend omits it", function()
    local backend = make_backend({})
    local info, err = Client.new(backend):hello()
    assert_nil(info)
    assert_eq(err.kind, "unsupported")
    assert_true(tostring(err):find("mock", 1, true) ~= nil, "unsupported error should name the transport")
  end)

  -- ── put / put_null ────────────────────────────────────────────────────────

  suite("Client:put")

  test("put forwards key, data, ttl, path", function()
    local seen
    local backend = make_backend({
      put = function(_, key, data, ttl, path) seen = { key = key, data = data, ttl = ttl, path = path }; return { ok = true } end,
    })
    local ok, err = Client.new(backend):put("mykey", { color = "blue" }, "60s", "/p")
    assert_nil(err)
    assert_true(ok)
    assert_eq(seen.key, "mykey")
    assert_eq(seen.data.color, "blue")
    assert_eq(seen.ttl, "60s")
    assert_eq(seen.path, "/p")
  end)

  test("put returns false+Error on backend error", function()
    local backend = make_backend({
      put = function() return { ok = false, error = { kind = "server_error", message = "must be a JSON object" } } end,
    })
    local ok, err = Client.new(backend):put("mykey", "not-an-object")
    assert_eq(ok, false)
    assert_true(tostring(err):find("object", 1, true) ~= nil)
  end)

  suite("Client:put_null")

  test("put_null forwards key and path", function()
    local seen
    local backend = make_backend({
      put_null = function(_, key, path) seen = { key = key, path = path }; return { ok = true } end,
    })
    local ok, err = Client.new(backend):put_null("mykey", "/somepath")
    assert_nil(err)
    assert_true(ok)
    assert_eq(seen.key, "mykey")
    assert_eq(seen.path, "/somepath")
  end)

  -- ── introspect ────────────────────────────────────────────────────────────

  suite("Client:introspect")

  test("introspect daemon returns typed daemon shape", function()
    local backend = make_backend({
      introspect = function(_, subject) return { ok = true, data = { pid = 1234, version = "0.9.0" } } end,
    })
    local result, err = Client.new(backend):introspect("daemon")
    assert_nil(err)
    assert_eq(result.subject, "daemon")
    assert_eq(result.daemon.pid, 1234)
    assert_nil(result.other)
  end)

  test("introspect non-daemon subject returns other", function()
    local backend = make_backend({
      introspect = function() return { ok = true, data = { entries = 5 } } end,
    })
    local result, err = Client.new(backend):introspect("cache")
    assert_nil(err)
    assert_eq(result.subject, "cache")
    assert_nil(result.daemon)
    assert_eq(result.other.entries, 5)
  end)

  test("introspect forwards subject and duration_secs", function()
    local seen
    local backend = make_backend({
      introspect = function(_, subject, duration_secs) seen = { subject = subject, duration_secs = duration_secs }; return { ok = true, data = {} } end,
    })
    Client.new(backend):introspect("timers", 30)
    assert_eq(seen.subject, "timers")
    assert_eq(seen.duration_secs, 30)
  end)

  test("introspect returns Error.unsupported when the backend omits it", function()
    local result, err = Client.new(make_backend({})):introspect("daemon")
    assert_nil(result)
    assert_eq(err.kind, "unsupported")
  end)

  -- ── resolve / eval ────────────────────────────────────────────────────────

  suite("Client:resolve")

  test("resolve forwards key, cwd, env, virtual overrides", function()
    local seen
    local backend = make_backend({
      resolve = function(_, key, cwd, env, overrides)
        seen = { key = key, cwd = cwd, env = env, overrides = overrides }
        return { ok = true, data = "baz" }
      end,
    })
    local value, err = Client.new(backend):resolve("filters.based", {
      cwd = "/tmp", env = { PYVAR = "/foo/bar/baz" }, virtual = { ["filters.based"] = "env.PYVAR | basename" },
    })
    assert_nil(err)
    assert_eq(value, "baz")
    assert_eq(seen.key, "filters.based")
    assert_eq(seen.cwd, "/tmp")
    assert_eq(seen.env.PYVAR, "/foo/bar/baz")
    assert_eq(seen.overrides["filters.based"], "env.PYVAR | basename")
  end)

  test("resolve defaults cwd to the context path", function()
    local seen_cwd
    local backend = make_backend({
      resolve = function(_, key, cwd) seen_cwd = cwd; return { ok = true, data = nil } end,
    })
    local c = Client.new(backend)
    c:set_context("/ctx")
    c:resolve("myproject")
    assert_eq(seen_cwd, "/ctx")
  end)

  test("resolve returns Error.unsupported when the backend omits it", function()
    local value, err = Client.new(make_backend({})):resolve("x.y")
    assert_nil(value)
    assert_eq(err.kind, "unsupported")
  end)

  suite("Client:eval")

  test("eval forwards template, cwd, env, overrides", function()
    local seen
    local backend = make_backend({
      eval = function(_, template, cwd, env, overrides)
        seen = { template = template, cwd = cwd }
        return { ok = true, data = "rendered" }
      end,
    })
    local value = Client.new(backend):eval("{{ git.branch }}", { cwd = "/repo" })
    assert_eq(value, "rendered")
    assert_eq(seen.template, "{{ git.branch }}")
    assert_eq(seen.cwd, "/repo")
  end)

  -- ── watch / WatchStream ───────────────────────────────────────────────────

  suite("Client:watch and WatchStream")

  test("watch opens via the backend and returns a WatchStream", function()
    local opened_with
    local watch_handle = {
      next_event = function(_, timeout_ms) return { data = 42, age_ms = 5, stale = false }, nil, "event" end,
      cancel = function(_) end,
      close = function(_) end,
    }
    local backend = make_backend({
      watch_open = function(_, key, path) opened_with = { key = key, path = path }; return watch_handle end,
    })
    local stream, err = Client.new(backend):watch("fixture.count", "/repo")
    assert_nil(err)
    assert_not_nil(stream)
    assert_eq(opened_with.key, "fixture.count")
    assert_eq(opened_with.path, "/repo")

    local ev = stream:next_event()
    assert_eq(ev.data, 42)
    assert_eq(ev.age_ms, 5)
  end)

  test("WatchStream:next_event returns nil, Error on outcome=error", function()
    local Error = require("beachcomber.error")
    local watch_handle = {
      next_event = function() return nil, Error.new("server_error", "watch closed"), "error" end,
    }
    local backend = make_backend({ watch_open = function() return watch_handle end })
    local stream = Client.new(backend):watch("k")
    local ev, err = stream:next_event()
    assert_nil(ev)
    assert_not_nil(err)
  end)

  test("WatchStream:next_event returns nil (no error) on outcome=eof", function()
    local watch_handle = { next_event = function() return nil, nil, "eof" end }
    local backend = make_backend({ watch_open = function() return watch_handle end })
    local stream = Client.new(backend):watch("k")
    local ev, err = stream:next_event()
    assert_nil(ev)
    assert_nil(err)
  end)

  test("WatchStream:each iterates events until eof", function()
    local calls = 0
    local watch_handle = {
      next_event = function()
        calls = calls + 1
        if calls <= 2 then return { data = calls }, nil, "event" end
        return nil, nil, "eof"
      end,
    }
    local backend = make_backend({ watch_open = function() return watch_handle end })
    local stream = Client.new(backend):watch("k")
    local collected = {}
    for ev in stream:each() do collected[#collected + 1] = ev.data end
    assert_eq(#collected, 2)
    assert_eq(collected[1], 1)
    assert_eq(collected[2], 2)
  end)

  test("watch returns Error.unsupported when the backend omits watch_open", function()
    local stream, err = Client.new(make_backend({})):watch("k")
    assert_nil(stream)
    assert_eq(err.kind, "unsupported")
  end)
end

File: sdks/ruby/test/test_ops.rb
require_relative 'test_helper'
require 'stringio'

class TestOps < Minitest::Test
  def setup
    @server = MockServer.new
    @client = Beachcomber::Client.new(socket_path: @server.path)
  end

  def teardown
    @server.stop
  end

  # --- hello ---

  def test_hello_sends_correct_op
    @server.enqueue('{"ok":true,"data":{"protocol_version":"1","daemon_version":"0.5.0"}}')
    @client.hello
    assert_equal 'hello', @server.requests.last['op']
  end

  def test_hello_returns_hello_info
    @server.enqueue('{"ok":true,"data":{"protocol_version":"1","daemon_version":"0.5.0"}}')
    info = @client.hello
    assert_instance_of Beachcomber::HelloInfo, info
    assert_equal '1',     info.protocol_version
    assert_equal '0.5.0', info.daemon_version
  end

  def test_hello_tolerates_missing_data_fields
    @server.enqueue('{"ok":true,"data":{}}')
    info = @client.hello
    assert_equal '', info.protocol_version
    assert_equal '', info.daemon_version
  end

  # --- put ---

  def test_put_sends_correct_op
    @server.enqueue('{"ok":true}')
    @client.put('custom.value', 'hello')
    req = @server.requests.last
    assert_equal 'put',          req['op']
    assert_equal 'custom.value', req['key']
    assert_equal 'hello',        req['data']
  end

  def test_put_returns_nil
    @server.enqueue('{"ok":true}')
    result = @client.put('custom.value', 42)
    assert_nil result
  end

  def test_put_with_ttl
    @server.enqueue('{"ok":true}')
    @client.put('custom.value', 'x', ttl: 60)
    req = @server.requests.last
    assert_equal 60, req['ttl']
  end

  def test_put_with_path
    @server.enqueue('{"ok":true}')
    @client.put('custom.value', 'x', path: '/repo')
    req = @server.requests.last
    assert_equal '/repo', req['path']
  end

  def test_put_omits_data_when_nil
    @server.enqueue('{"ok":true}')
    @client.put('custom.value')
    req = @server.requests.last
    refute req.key?('data')
  end

  def test_put_omits_ttl_when_not_given
    @server.enqueue('{"ok":true}')
    @client.put('custom.value', 'x')
    req = @server.requests.last
    refute req.key?('ttl')
  end

  # --- introspect ---

  def test_introspect_sends_correct_op_and_subject
    @server.enqueue('{"ok":true,"data":{"provider_count":5}}')
    @client.introspect(Beachcomber::IntrospectSubject::PROVIDERS)
    req = @server.requests.last
    assert_equal 'introspect', req['op']
    assert_equal 'providers',  req['subject']
  end

  def test_introspect_returns_introspect_response
    @server.enqueue('{"ok":true,"data":{"provider_count":5}}')
    resp = @client.introspect('providers')
    assert_instance_of Beachcomber::IntrospectResponse, resp
    assert_equal 'providers', resp.subject
    assert_nil resp.daemon
    assert_equal({ 'provider_count' => 5 }, resp.other)
  end

  def test_introspect_daemon_subject_returns_daemon_health
    payload = {
      ok: true,
      data: {
        pid: 1234,
        version: '0.5.0',
        uptime_secs: 3600,
        socket_path: '/tmp/sock',
        config_path: '/home/user/.config/beachcomber/config.toml',
        requests_total: 100,
        in_flight: 2,
        active_watchers: 1,
        cache_entries: 42,
        verdicts: [{ level: 'ok', message: 'all good' }],
      }
    }.to_json
    @server.enqueue(payload)
    resp = @client.introspect(Beachcomber::IntrospectSubject::DAEMON)
    assert_instance_of Beachcomber::IntrospectResponse, resp
    assert_equal 'daemon', resp.subject
    assert_nil resp.other

    health = resp.daemon
    assert_instance_of Beachcomber::DaemonHealth, health
    assert_equal 1234,    health.pid
    assert_equal '0.5.0', health.version
    assert_equal 3600,    health.uptime_secs
    assert_equal '/tmp/sock', health.socket_path
    assert_equal 100,     health.requests_total
    assert_equal 2,       health.in_flight
    assert_equal 1,       health.active_watchers
    assert_equal 42,      health.cache_entries
    assert_equal 1,       health.verdicts.size
    assert_instance_of Beachcomber::Verdict, health.verdicts.first
    assert_equal 'ok',       health.verdicts.first.level
    assert_equal 'all good', health.verdicts.first.message
  end

  def test_introspect_with_duration_secs
    @server.enqueue('{"ok":true,"data":null}')
    @client.introspect('cache', duration_secs: 5)
    req = @server.requests.last
    assert_equal 5, req['duration_secs']
  end

  def test_introspect_omits_duration_when_not_given
    @server.enqueue('{"ok":true,"data":null}')
    @client.introspect('cache')
    req = @server.requests.last
    refute req.key?('duration_secs')
  end

  # --- status ---

  def test_status_sends_status_op
    @server.enqueue('{"ok":true,"data":[]}')
    @client.status
    assert_equal 'status', @server.requests.last['op']
  end

  def test_status_returns_array_of_cache_rows
    payload = {
      ok: true,
      data: [
        { provider: 'git', field: 'branch', path: '/repo', value: 'main', age_ms: 100, stale: false },
        { provider: 'git', field: 'dirty',  path: '/repo', value: true,   age_ms: 200, stale: true  },
      ]
    }.to_json
    @server.enqueue(payload)
    rows = @client.status
    assert_equal 2, rows.size
    rows.each { |r| assert_instance_of Beachcomber::CacheRow, r }
    assert_equal 'git',   rows[0].provider
    assert_equal 'branch', rows[0].field
    assert_equal 'main',  rows[0].value
    assert_equal 100,     rows[0].age_ms
    refute rows[0].stale
    assert rows[1].stale
  end

  def test_status_raises_protocol_error_when_data_not_array
    @server.enqueue('{"ok":true,"data":{"not":"an array"}}')
    assert_raises(Beachcomber::ProtocolError) { @client.status }
  end

  # --- get_with_flags ---

  def test_get_with_flags_sends_force
    @server.enqueue('{"ok":true,"data":"main","age_ms":0,"stale":false}')
    @client.get_with_flags('git.branch', force: true)
    req = @server.requests.last
    assert_equal 'get', req['op']
    assert_equal true,  req['force']
    refute req.key?('wait')
  end

  def test_get_with_flags_sends_wait
    @server.enqueue('{"ok":true,"data":"main","age_ms":0,"stale":false}')
    @client.get_with_flags('git.branch', wait: true)
    req = @server.requests.last
    assert_equal true, req['wait']
    refute req.key?('force')
  end

  def test_get_with_flags_omits_flags_by_default
    @server.enqueue('{"ok":true,"data":"main","age_ms":0,"stale":false}')
    @client.get_with_flags('git.branch')
    req = @server.requests.last
    refute req.key?('force')
    refute req.key?('wait')
  end

  def test_get_with_flags_returns_result
    @server.enqueue('{"ok":true,"data":"main","age_ms":55,"stale":false}')
    result = @client.get_with_flags('git.branch')
    assert_instance_of Beachcomber::Result, result
    assert_equal 'main', result.data
    assert_equal 55,     result.age_ms
  end

  # --- watch ---

  def test_watch_sends_watch_op
    # The watch stream stays open; we use the MockServer's streaming behaviour.
    # Build a minimal response then close the stream immediately.
    @server.enqueue('{"ok":true,"data":"main","age_ms":10,"stale":false}')
    stream = @client.watch('git.branch')
    event = stream.next_event
    stream.close

    req = @server.requests.last
    assert_equal 'watch',      req['op']
    assert_equal 'git.branch', req['key']
    assert_instance_of Beachcomber::WatchEvent, event
    assert_equal 'main', event.data
    assert_equal 10,     event.age_ms
    refute event.stale
  end

  def test_watch_with_path_includes_path
    @server.enqueue('{"ok":true,"data":"main","age_ms":0,"stale":false}')
    stream = @client.watch('git.branch', path: '/repo')
    stream.next_event
    stream.close
    req = @server.requests.last
    assert_equal '/repo', req['path']
  end

  def test_watch_stream_is_enumerable
    assert_includes Beachcomber::WatchStream.ancestors, Enumerable
  end

  def test_watch_stream_each_yields_watch_events
    # Build a fake socket that delivers 3 watch events then EOF.
    lines = 3.times.map { '{"ok":true,"data":"val","age_ms":1,"stale":false}' }.join("\n") + "\n"
    fake_io = StringIO.new(lines)
    stream  = Beachcomber::WatchStream.new(fake_io)
    events  = stream.to_a
    assert_equal 3, events.size
    events.each { |e| assert_instance_of Beachcomber::WatchEvent, e }
  end

  def test_watch_event_stale_flag
    @server.enqueue('{"ok":true,"data":"old","age_ms":9000,"stale":true}')
    stream = @client.watch('git.branch')
    event = stream.next_event
    stream.close
    assert event.stale
  end

  # --- session additions ---

  def test_session_hello
    @server.enqueue('{"ok":true,"data":{"protocol_version":"2","daemon_version":"1.0.0"}}')
    result = nil
    @client.session do |s|
      result = s.hello
    end
    assert_instance_of Beachcomber::HelloInfo, result
    assert_equal '2',     result.protocol_version
    assert_equal '1.0.0', result.daemon_version
  end

  def test_session_put
    @server.enqueue('{"ok":true}')
    @client.session do |s|
      result = s.put('mykey', 'myvalue')
      assert_nil result
    end
    req = @server.requests.last
    assert_equal 'put',     req['op']
    assert_equal 'mykey',   req['key']
    assert_equal 'myvalue', req['data']
  end

  def test_session_introspect
    @server.enqueue('{"ok":true,"data":{"items":[]}}')
    result = nil
    @client.session do |s|
      result = s.introspect('cache')
    end
    assert_instance_of Beachcomber::IntrospectResponse, result
    assert_equal 'cache', result.subject
  end

  def test_session_status
    @server.enqueue('{"ok":true,"data":[{"provider":"git","field":"branch","path":null,"value":"main","age_ms":50,"stale":false}]}')
    rows = nil
    @client.session do |s|
      rows = s.status
    end
    assert_equal 1, rows.size
    assert_instance_of Beachcomber::CacheRow, rows.first
  end

  def test_session_get_with_flags
    @server.enqueue('{"ok":true,"data":"main","age_ms":0,"stale":false}')
    result = nil
    @client.session do |s|
      result = s.get_with_flags('git.branch', force: true)
    end
    assert_instance_of Beachcomber::Result, result
    assert_equal 'main', result.data
    req = @server.requests.last
    assert_equal true, req['force']
  end

  # --- lifecycle fields on CacheRow ---

  def test_status_row_exposes_lifecycle_fields
    payload = {
      ok: true,
      data: [
        {
          provider: 'git',
          field: 'branch',
          path: '/repo',
          value: 'main',
          age_ms: 100,
          stale: false,
          kind: { kind: 'lifecycle', decay: 0, watches_files: true },
          poll_interval_secs: 30,
          keep_alive_polls: 3,
          fsevents_reinstate: false,
        }
      ]
    }.to_json
    @server.enqueue(payload)
    rows = @client.status
    git = rows.find { |r| r.provider == 'git' }
    refute_nil git.kind
    assert_equal 'lifecycle', git.kind['kind']
    assert git.poll_interval_secs > 0
    assert git.keep_alive_polls > 0
    refute_nil git.fsevents_reinstate
  end

  # --- IntrospectSubject constants ---

  def test_introspect_subject_constants_defined
    assert_equal 'daemon',    Beachcomber::IntrospectSubject::DAEMON
    assert_equal 'providers', Beachcomber::IntrospectSubject::PROVIDERS
    assert_equal 'config',    Beachcomber::IntrospectSubject::CONFIG
    assert_equal 'cache',     Beachcomber::IntrospectSubject::CACHE
    assert_equal 'lifecycle', Beachcomber::IntrospectSubject::LIFECYCLE
    assert_equal 'watches',   Beachcomber::IntrospectSubject::WATCHES
    assert_equal 'timers',    Beachcomber::IntrospectSubject::TIMERS
    assert_equal 'demand',    Beachcomber::IntrospectSubject::DEMAND
    assert_equal 'procs',     Beachcomber::IntrospectSubject::PROCS
  end
end
